ark/phenotyping/som_utils.py

```diff
@@ -80,46 +83,96 @@ def create_pixel_matrix(img_xr, seg_labels, fovs=None, channels=None, blur_facto
+        # write dataset to hdf5 file, mode 'a' will create if it doesn't exist
+        pixel_mat.to_hdf(os.path.join(base_dir, hdf_name), key=fov, mode='a')
+
+
+def subset_pixels(fovs, base_dir, hdf_name='pixel_mat_preprocessed.hdf5',
+                  csv_name='pixel_mat_subsetted.csv', subset_percent=0.1):
+    """Takes a random percentage subset of pixel data from each fov
+
+    Args:
+        fovs (list):
+            The list of fovs to read
+        base_dir (str):
+            Name of the directory to save the subsetted CSV to
+        hdf_name (str):
+            Name of the file which contains the preprocessed pixel data,
+            defaults to pixel_mat_preprocessed.hdf5
+        csv_name (str):
+            Name of the file to write the subsetted pixel data frame to
+        subset_percent (float):
+            The percentage of pixels to take from each fov, defaults to 0.1
+    """
+
+    # ensure subset percent is valid
+    if subset_percent <= 0 or subset_percent > 1:
+        raise ValueError('Subset percent provided must be in (0, 1]')
+
+    # ensure the file path to hdf_name exists
+    if not os.path.exists(os.path.join(base_dir, hdf_name)):
+        raise FileNotFoundError('Preprocessed HDF5 %s not found in base_dir %s' %
+                                (hdf_name, base_dir))
 
-    # normalize each marker column by the 99.9 percentile value
-    flowsom_data.loc[:, channels] = flowsom_data.loc[:, channels].div(
-        flowsom_data.loc[:, channels].quantile(q=0.999, axis=0), axis=1)
+    # define our subsetted flowsom matrix
+    flowsom_subset_data = None
 
-    return flowsom_data
+    for fov in fovs:
+        # read the specific fov key from the HDF5
+        fov_pixel_data = pd.read_hdf(os.path.join(base_dir, hdf_name), key=fov)
+
+        # subset the data per fov using the subset_percent argument
+        fov_pixel_data = fov_pixel_data.sample(frac=subset_percent)
 
+        # assign to flowsom_subset_data if not already assigned, otherwise concatenates
+        if flowsom_subset_data is None:
+            flowsom_subset_data = fov_pixel_data
+        else:
+            flowsom_subset_data = pd.concat([flowsom_subset_data, fov_pixel_data])
 
-def cluster_pixels(chan_list, base_dir,
-                   pixel_pre_name='pixel_mat_preprocessed.csv',
-                   pixel_cluster_name='pixel_mat_clustered.csv'):
+    flowsom_subset_data.to_csv(os.path.join(base_dir, csv_name), index=False)
+
+
+def cluster_pixels(fovs, channels, base_dir,
+                   pixel_pre_name='pixel_mat_preprocessed.hdf5',
+                   pixel_subset_name='pixel_mat_subsetted.csv',
+                   pixel_cluster_name='pixel_mat_clustered.hdf5'):
     """Run the FlowSOM training on the pixel data.
 
     Saves results to pixel_mat_clustered.csv in base_dir.
-    Usage: Rscript som_runner.R {path_to_pixel_matrix} {chan_list_comma_separated} {save_path}
+    Usage: Rscript som_runner.R {fovs} {chan_list} {pixel_matrix_path}
```

Comment:
This usage description isn't actually for this function though, it's for the R script, right?

Reply:
Yes, I guess we don't need to include this since we wouldn't expect the user to have to run the R script directly.